src/engine/controls/macrocontrol.cpp

```diff
@@ -0,0 +1,262 @@
+#include "macrocontrol.h"
+
+namespace {
+constexpr uint kRecordingTimerInterval = 100;
+constexpr size_t kRecordingQueueSize = kRecordingTimerInterval / 10;
+} // namespace
+
+ConfigKey MacroControl::getConfigKey(QString name) {
+    return ConfigKey(m_group, m_controlPattern.arg(name));
+}
+
+MacroControl::MacroControl(QString group, UserSettingsPointer pConfig, int slot)
+        : EngineControl(group, pConfig),
+          m_slot(slot),
+          m_controlPattern(QString("macro_%1_%2").arg(slot)),
+          m_bJumpPending(false),
+          m_recordedActions(kRecordingQueueSize),
+          m_iNextAction(0),
+          m_COPlaying(getConfigKey("playing")),
+          m_COStatus(getConfigKey("status")),
+          m_record(getConfigKey("record")),
```

Comment:
The same here, this should also become a toggle button.

Reply:
Yeah, seems sensible. I wasn't quite sure what a toggle button really does.
I thought it wouldn't work if the state can be changed by other controls.

Reply:
A full featured prototype you can look at is the normal play button.
It toggles and gives a feedback in turn.  I think something like this will work here as well. 
You can even toggle through more than one state.
This can be sometimes tricky, see the broadcasting feature.

Reply:
Fixed in https://github.com/mixxxdj/mixxx/pull/2989/commits/4a03821e43113b238e3ffe2f99384b4468a0adfb